uktrade/enquiry-mgmt-tool
# Change the way DataHub metadata are fetched, cached and error handled

## Description of change

This PR fixes the problematic fetching, caching and error handling of DataHub metadata and greatly simplifies it. Before this PR, the 10 metadata requests were all done at once (not in parallel though) in the [`dh_fetch_metadata`](https://github.com/uktrade/enquiry-mgmt-tool/blob/master/app/enquiries/common/datahub_utils.py#L134) and then all of them combined into a single `dict` which was subsequently cached from within the `dh_load_metadata` function. The failed requests were handled so, that a name of the failed metadata was added to a `list` in the `'failed'` key of the resulting `dict`, which broke the code which was using the `dh_fetch_metadata` for the duration of the expiration time of the cached `dict`. The result was that the real error for the _submit to DataHub_ functionality was not working was never reported.

This PR replaces the `dh_fetch_metadata` function with a `fetch_metadata` function which only fetches a single metadata endpoint at a time. The function is then called on demand where the given metadata is needed. If the underlying reques fails, the function raises an exception with [`response.raise_for_status()`](https://requests.readthedocs.io/en/master/api/#requests.Response.raise_for_status).

## Test instructions

1. Open an enquiry which can be sumbitted to DataHub
2. Click the _Submit to DataHub_ button
3. Correct all the eventual data related errors
4. The enquiry should finally be succesfully submitted to DataHub
5. Open your `.env` file and either change the `DATA_HUB_METADATA_URL` so that it's invalid, or alternatively tweak one of the `DATA_HUB_HAWK_ID` or `DATA_HUB_HAWK_KEY` values to make the metadata requests fail.
6. Restart the application so that the changes take effect
7. Just to be sure, clear the Django cache with `docker-compose exec  app python manage.py clear_cache`
8. Repeat steps 1 to 4
9. The application should render a `404` or `401` error, depending on the way you made the metadata requests fail.
10. The error related to the failed request should appear in the log.
